<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Sutherland project preliminary findings</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="report_files/libs/clipboard/clipboard.min.js"></script>
<script src="report_files/libs/quarto-html/quarto.js"></script>
<script src="report_files/libs/quarto-html/popper.min.js"></script>
<script src="report_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="report_files/libs/quarto-html/anchor.min.js"></script>
<link href="report_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="report_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="report_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="report_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="report_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Sutherland project preliminary findings</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="preprocessing-workflow" class="level2">
<h2 class="anchored" data-anchor-id="preprocessing-workflow">Preprocessing workflow</h2>
<p>Each sample underwent ambient RNA correction (SoupX), red blood cell filtering (HBA1, HBA2, HBB greater than 5% of total counts), damaged cell filtering (DamageDetective, v2.0.1), and doublet removal (DoubletFinder). Counts were scaled and normalised before integration using canonical correlation analysis on the top 30 principal components and 2000 most variable genes (Seurat).</p>
</section>
<section id="sample-information" class="level2">
<h2 class="anchored" data-anchor-id="sample-information">Sample information</h2>
<p><strong>Table</strong> Sample summary statistics following pre-processing</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Sample ID</th>
<th style="text-align: left;">BAL Cell Number</th>
<th style="text-align: left;">PBMC Cell Number</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ID_1098</td>
<td style="text-align: left;">667</td>
<td style="text-align: left;">575</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID_1376</td>
<td style="text-align: left;">1491 *</td>
<td style="text-align: left;">530</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID_1483</td>
<td style="text-align: left;">731 *</td>
<td style="text-align: left;">66</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID_1484</td>
<td style="text-align: left;">-</td>
<td style="text-align: left;">260</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID_1523</td>
<td style="text-align: left;">203</td>
<td style="text-align: left;">203</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID_1566</td>
<td style="text-align: left;">791 *</td>
<td style="text-align: left;">209</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID_1676</td>
<td style="text-align: left;">250 *</td>
<td style="text-align: left;">86</td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<p>Note: BAL samples showed unusual clustering patterns in samples *<a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/BAL_1376.png">BAL_1376</a>, *<a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/BAL_1483.png">BAL_1483</a>, *<a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/BAL_1566.png">BAL_1566</a> and, to a small extent, *<a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/BAL_1676.png">BAL_1676</a>. This is typical of low quality samples. This is clear when comparing the clustering to high quality samples, <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/BAL_1098.png">BAL_1098</a> and <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/BAL_1523.png">BAL_1523</a>.</p>
<p>All PBMC samples were of a high quality <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1098.png">PBMC_1098</a>, <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1376.png">PBMC_1376</a>, <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1483.png">PBMC_1483</a>, <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1484.png">PBMC_1484</a>, <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1523.png">PBMC_1523</a>, <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1566.png">PBMC_1566</a>, and <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/sample_quality/PBMC_1676.png">PBMC_1676</a>.</p>
</blockquote>
</section>
<section id="visualisations" class="level2">
<h2 class="anchored" data-anchor-id="visualisations">Visualisations</h2>
<p>The integrated dataset is shown below across all samples coloured according to cell types.</p>
<p><strong>Lymphoid cells</strong> formed clear clusters from PBMC samples that included T and B cells, with NK cells collecting with T cells. Myeloid populations originating from PBMCs are collected into a monocyte cluster made of distinct classical (cMono) and non-classical (ncMono) subsets. BAL samples contained predominantly myeloid cells, which, due to low sample quality, showed distinct <a href="https://github.com/AlicenJoyHenning/TB_BAL/blob/master/plots/integrated/batch_effect.png">batch effects</a> despite robust preprocessing. For this reason, the myeloid clusters cannot be reliably annotated to finer labels, such as into macrophages or dendritic cells, despite it being likely that they are present.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="plots/integrated/annotated.svg" class="img-fluid figure-img"></p>
<figcaption><strong>Figure</strong> UMAP of annotated samples from the integrated TB BAL-PBMC dataset</figcaption>
</figure>
</div>
</section>
<section id="composition" class="level2">
<h2 class="anchored" data-anchor-id="composition">Composition</h2>
<p>Cell composition varied across samples, some with a heavy myeloid composition and others with a heavy lymphoid population. However, when viewing the compositions according to tissue origin, we see this is most likely due to differences in BAL and PBMC proportions of the corresponding samples.</p>
<p>In general, samples of BAL origin are rich in myeloid cells while PBMCs are rich in lymphoid cells. But even in accounting for sample origin, there is dissimilarity in cell composition between samples.</p>
<section id="total-cell-compositions" class="level4">
<h4 class="anchored" data-anchor-id="total-cell-compositions">Total cell compositions</h4>
<p><img src="plots/composition/PBMC_BAL_proportion.svg" class="img-fluid"></p>
<p><strong>Figure</strong> Proportion of each cell type within samples of both BAL and PBMC origin</p>
<p><strong>Table</strong> Actual cell numbers originating from each sample of both BAL and PBMC origin</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Sample</th>
<th style="text-align: right;">B</th>
<th style="text-align: right;">cMono</th>
<th style="text-align: right;">myeloid</th>
<th style="text-align: right;">ncMono</th>
<th style="text-align: right;">T</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ID_1098</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">1013</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">182</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID_1376</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">1502</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">452</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID_1483</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">717</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">49</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID_1484</td>
<td style="text-align: right;">55</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">168</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID_1523</td>
<td style="text-align: right;">80</td>
<td style="text-align: right;">72</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">64</td>
<td style="text-align: right;">188</td>
</tr>
<tr class="even">
<td style="text-align: left;">ID_1566</td>
<td style="text-align: right;">61</td>
<td style="text-align: right;">32</td>
<td style="text-align: right;">794</td>
<td style="text-align: right;">31</td>
<td style="text-align: right;">82</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ID_1676</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">241</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">49</td>
</tr>
</tbody>
</table>
</section>
<section id="isolated-bal-and-pbmc-compositions" class="level4">
<h4 class="anchored" data-anchor-id="isolated-bal-and-pbmc-compositions">Isolated BAL and PBMC compositions</h4>
<p><strong>BAL</strong></p>
<p><img src="plots/composition/BAL_proportion.svg" class="img-fluid"></p>
<p><strong>Figure</strong> Cell compositions of BAL sample origin</p>
<p><strong>Table</strong> Actual cell numbers for BAL sample origin</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: center;">Sample</th>
<th style="text-align: center;">B</th>
<th style="text-align: center;">T</th>
<th style="text-align: center;">cMono</th>
<th style="text-align: center;">ncMono</th>
<th style="text-align: center;">myeloid</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">ID_1098</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">65</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">600</td>
</tr>
<tr class="even">
<td style="text-align: center;">ID_1376</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">1491</td>
</tr>
<tr class="odd">
<td style="text-align: center;">ID_1483</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">13</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">717</td>
</tr>
<tr class="even">
<td style="text-align: center;">ID_1523</td>
<td style="text-align: center;">40</td>
<td style="text-align: center;">94</td>
<td style="text-align: center;">36</td>
<td style="text-align: center;">32</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">ID_1566</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">12</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">778</td>
</tr>
<tr class="even">
<td style="text-align: center;">ID_1676</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">9</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">241</td>
</tr>
</tbody>
</table>
<p><strong>PBMC</strong></p>
<p><img src="plots/composition/PBMC_proportion.svg" class="img-fluid"></p>
<p><strong>Figure</strong> Cell compositions of PBMC sample origin</p>
<p><strong>Table</strong> Actual cell numbers for PBMC sample origin</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: center;">Sample</th>
<th style="text-align: center;">B</th>
<th style="text-align: center;">T</th>
<th style="text-align: center;">cMono</th>
<th style="text-align: center;">ncMono</th>
<th style="text-align: center;">myeloid</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">ID_1098</td>
<td style="text-align: center;">22</td>
<td style="text-align: center;">117</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">413</td>
</tr>
<tr class="even">
<td style="text-align: center;">ID_1376</td>
<td style="text-align: center;">36</td>
<td style="text-align: center;">452</td>
<td style="text-align: center;">13</td>
<td style="text-align: center;">18</td>
<td style="text-align: center;">11</td>
</tr>
<tr class="odd">
<td style="text-align: center;">ID_1483</td>
<td style="text-align: center;">23</td>
<td style="text-align: center;">36</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">NA</td>
</tr>
<tr class="even">
<td style="text-align: center;">ID_1484</td>
<td style="text-align: center;">55</td>
<td style="text-align: center;">168</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">21</td>
<td style="text-align: center;">8</td>
</tr>
<tr class="odd">
<td style="text-align: center;">ID_1523</td>
<td style="text-align: center;">40</td>
<td style="text-align: center;">94</td>
<td style="text-align: center;">36</td>
<td style="text-align: center;">32</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">ID_1566</td>
<td style="text-align: center;">60</td>
<td style="text-align: center;">70</td>
<td style="text-align: center;">32</td>
<td style="text-align: center;">31</td>
<td style="text-align: center;">16</td>
</tr>
<tr class="odd">
<td style="text-align: center;">ID_1676</td>
<td style="text-align: center;">44</td>
<td style="text-align: center;">40</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<ul>
<li><p>Young, M.D., Behjati, S. (2020). SoupX removes ambient RNA contamination from droplet-based single-cell RNA sequencing data, GigaScience, Volume 9, Issue 12, December 2020, giaa151bioRxiv, 303727, <a href="https://doi.org/10.1093/gigascience/giaa151" class="uri">https://doi.org/10.1093/gigascience/giaa151</a></p></li>
<li><p>Hao, Y., Stuart, T., Kowalski, M., Choudhary, S., Hoffman, P., Hartman, A., Srivastava, A., Molla, G., Madad, S., Fernandez-Granda, C. &amp; Satija, R. (2024) ‘Dictionary learning for integrative, multimodal and massively scalable single-cell analysis’, <em>Nature Biotechnology</em>, 42(2), pp.&nbsp;293–304. <a href="doi:10.1038/s41587-023-01767-y" class="uri">doi:10.1038/s41587-023-01767-y</a>.</p></li>
<li><p>McGinnis, C.S., Murrow, L.M. and Gartner, Z.J. (2019). DoubletFinder: Doublet detection in single-cell RNA sequencing data using artificial nearest neighbors. <em>Cell Systems</em>, 8(4), pp.329-337.e4. <a href="https://doi.org/10.1016/j.cels.2019.03.003" class="uri">https://doi.org/10.1016/j.cels.2019.03.003</a></p></li>
</ul>
<p>Note this workflow made use of the DamageDetective R package we are in the process of writing up that is available on <a href="https://github.com/AlicenJoyHenning/DamageDetective">GitHub</a> and <a href="https://cloud.r-project.org/web/packages/DamageDetective/index.html">CRAN</a>.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>